File: leet-code/393. UTF-8 Validation/UTF8Validation393_2.java
package com.example;

public class UTF8Validation393_2 {
    public boolean validUtf8(int[] data) {
        int idx = 0;
        while (idx < data.length) {
            int num = data[idx];
            int firstBitCount = getFirstBitCount(num);
            if (firstBitCount == 1 || firstBitCount > 4 || (firstBitCount > 0 && idx + firstBitCount > data.length)) {
                return false;
            }
            for (int i = idx + 1; i < idx + firstBitCount; i++) {
                int bitCount = getFirstBitCount(data[i]);
                if (bitCount != 1) {
                    return false;
                }
            }
            if (firstBitCount == 0) {
                idx += 1;
            } else {
                idx += firstBitCount;
            }
        }
        return true;
    }

    private int getFirstBitCount(int num) {
        // 8자리 2진수로 만들어서 확인해야 한다
        String binary = String.format("%8s", Integer.toBinaryString(num))
                .replace(" ", "0");
        int count = 0;
        for (char b : binary.toCharArray()) {
            if (b == '1') {
                count += 1;
            } else {
                break;
            }
        }
        return count;
    }
}

File: leet-code/393. UTF-8 Validation/UTF8Validation393_2.kt
package com.example

class UTF8Validation393_2 {
    fun validUtf8(data: IntArray): Boolean {
        var idx = 0
        while (idx < data.size) {
            idx += when {
                data[idx] shr 3 == 0b11110 && idx + 3 < data.size && isValid(data, idx + 1, idx + 3) -> 4
                data[idx] shr 4 == 0b1110 && idx + 2 < data.size && isValid(data, idx + 1, idx + 2) -> 3
                data[idx] shr 5 == 0b110 && idx + 1 < data.size && isValid(data, idx + 1, idx + 1) -> 2
                data[idx] shr 6 == 0b10 -> return false
                data[idx] shr 7 == 0 -> 1
                else -> return false
            }
        }
        return true
    }

    private fun isValid(data: IntArray, start: Int, end: Int): Boolean {
        for (i in start..end) {
            if (data[i] shr 6 != 0b10) {
                return false
            }
        }
        return true
    }
}

// 시프트 연산 참고
// https://codedragon.tistory.com/7998
